internal/ssc/ssc_proxy.go

```diff
@@ -0,0 +1,226 @@
+	"github.com/sourcegraph/sourcegraph/schema"
+)
+
+// SSCAPIProxy is an HTTP handler that essentially proxies API requests from the
+// current Sourcegraph instance to the SSC backend, but exchanging the credentials
+// of the calling Sourcegraph user with an access token for their SAMS identity.
+//
+// This way we can transparently serve an HTTP API from Sourcegraph.com, when in
+// actuality it is processed by a different service all together. (And thereby
+// allowing us to decouple Cody Pro-specifc functionality from the Sourcegraph
+// Enterprise instance.)
+//
+// The API Proxy simply handles the credential exchange and verification. But
+// will send along the HTTP request. (Even if the URL and method aren't supported,
+// and would serve a 404/405 response, etc.)
+type APIProxyHandler struct {
+	CodyProConfig *schema.CodyProConfig
+	DB            database.DB
+	Logger        log.Logger
+
+	// URLPrefix of where the handler is served. e.g. ".api/ssc/proxy/". This
+	// will be replaced with the SSC-specific URL prefix ()"cody/api/v1/").
+	URLPrefix string
+}
+
+var _ http.Handler = (*APIProxyHandler)(nil)
+
+// getUserIDFromRequest extracts the Sourcegraph User ID from the incomming request,
+// or returns an error suitable for sending to the end user.
+func (p *APIProxyHandler) getUserIDFromContext(ctx context.Context) (int32, error) {
+	callingActor := actor.FromContext(ctx)
+	if callingActor == nil || !callingActor.IsAuthenticated() {
+		p.Logger.Warn("rejecting request made by unauthenticated Sourcegraph user")
+		return 0, errors.New("no credentials available")
+	}
+	if callingActor.IsInternal() || callingActor.SourcegraphOperator {
+		p.Logger.Warn("rejecting request made by internal service / Sourcegraph Operator")
+		return 0, errors.New("request not made on behalf of a user")
+	}
+	return callingActor.UID, nil
+}
+
+// buildProxyRequest converts the incomming HTTP request into what will be sent to the SSC backend.
+func (p *APIProxyHandler) buildProxyRequest(sourceReq *http.Request, token string) (*http.Request, error) {
+	// For simplicity, read the full request body before sending the proxy request.
+	var bodyReader io.Reader
+	if sourceReq.Body != nil {
+		bodyBytes, err := io.ReadAll(sourceReq.Body)
+		if err != nil {
+			return nil, errors.Wrap(err, "reading source request body")
+		}
+		bodyReader = strings.NewReader(string(bodyBytes))
+	}
+
+	// Construct the remapped URL on the SSC backend.
+	// For example:
+	// Source: ".api/ssc/proxy/" + "teams/current/members"
+	// Proxy :    "cody/api/v1/" + "teams/current/members"
+	sourceURLPath := strings.TrimPrefix(sourceReq.URL.Path, p.URLPrefix)
+	sscURL := fmt.Sprintf(
+		"%s/cody/api/v1/%s?%s",
+		p.CodyProConfig.SscBackendOrigin,
+		sourceURLPath,
+		sourceReq.URL.Query().Encode())
+	p.Logger.Info("Building proxy request", log.String("method", sourceReq.Method), log.String("url", sscURL))
+	proxyReq, err := http.NewRequest(sourceReq.Method, sscURL, bodyReader)
```

Comment:
I would recommend using `url.JoinPath` here, instead of using `fmt.Sprintf`, and call `path.Clean` before joining the API path with the user controlled bits. This makes sure we always go to `cody/api/v1`. 

```suggestion
	sourceURLPath := strings.TrimPrefix(sourceReq.URL.Path, p.URLPrefix)
	// make sure the path is clean
	apiPath := fmt.Sprintf("/cody/api/v1/%s", path.Clean(sourceURLPath))
	
	fullURL, err := url.JoinPath(baseURL.String(), apiPath)
	if err != nil {
		return nil, errors.Wrap(err, "error constructing fullpath")
	}

	sscURL, err := url.Parse(fullURL)
	if err != nil {
		return "", errors.Wrap(err, "error parsing full path)
	}
	sscURL.RawQuery = sourceReq.Query().Encode()
	proxyReq, err := http.NewRequest(sourceReq.Method, sscURL.String(), bodyReader)
```